lib/charms/mongodb/v0/mongo.py

```diff
@@ -0,0 +1,284 @@
+"""Code for interactions with MongoDB."""
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+import logging
+import re
+from dataclasses import dataclass
+from itertools import chain
+from typing import List, Optional, Set
+from urllib.parse import quote_plus
+
+from pymongo import MongoClient
+from pymongo.errors import OperationFailure, PyMongoError
+from tenacity import RetryError, Retrying, stop_after_delay, wait_fixed
+
+from config import Config
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+class NotReadyError(PyMongoError):
+    """Raised when not all replica set members healthy or finished initial sync."""
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3037662a76cc4bf1876d4659c88e77e5"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+ADMIN_AUTH_SOURCE = "authSource=admin"
+SYSTEM_DBS = ("admin", "local", "config")
+REGULAR_ROLES = {
+    "admin": [
+        {"role": "userAdminAnyDatabase", "db": "admin"},
+        {"role": "readWriteAnyDatabase", "db": "admin"},
+        {"role": "userAdmin", "db": "admin"},
+    ],
+    "monitor": [
+        {"role": "explainRole", "db": "admin"},
+        {"role": "clusterMonitor", "db": "admin"},
+        {"role": "read", "db": "local"},
+    ],
+    "backup": [
+        {"db": "admin", "role": "readWrite", "collection": ""},
+        {"db": "admin", "role": "backup"},
+        {"db": "admin", "role": "clusterMonitor"},
+        {"db": "admin", "role": "restore"},
+        {"db": "admin", "role": "pbmAnyAction"},
+    ],
+}
+
+# path to store mongodb ketFile
+logger = logging.getLogger(__name__)
+
+
+class AmbigiousConfigError(Exception):
+    """Raised when the config could correspond to a mongod config or mongos config."""
+
+
+@dataclass
+class MongoConfiguration:
+    """Class for Mongo configurations usable my mongos and mongodb.
+
+    — replset: name of replica set
+    — database: database name.
+    — username: username.
+    — password: password.
+    — hosts: full list of hosts to connect to, needed for the URI.
+    — tls_external: indicator for use of internal TLS connection.
+    — tls_internal: indicator for use of external TLS connection.
+    """
+
+    database: Optional[str]
+    username: str
+    password: str
+    hosts: Set[str]
+    roles: Set[str]
+    tls_external: bool
+    tls_internal: bool
+    port: Optional[str] = Config.MONGODB_PORT
+    replset: Optional[str] = None
```

Comment:
Beware : port is an int not a str + it's always defined because of the default value
```suggestion
    port: int = Config.MONGODB_PORT
    replset: str | None = None
```